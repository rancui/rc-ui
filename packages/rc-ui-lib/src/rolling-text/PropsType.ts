import { BaseTypeProps } from '../utils';

export type RollingTextDirection = 'up' | 'down';
export type RollingTextStopOrder = 'ltr' | 'rtl';

export interface RollingTextProps extends BaseTypeProps {
  /**
   * 数字高度，单位为 px
   * @default 40
   */
  height?: number;
  /**
   * 起始数值
   * @default 0
   */
  startNum?: number;
  /**
   * 目标数值
   */
  targetNum?: number;
  /**
   * 内容数组，用于翻转非数字内容
   * @default []
   */
  textList?: string[];
  /**
   * 动画时长，单位为秒
   * @default 2
   */
  duration?: number;
  /**
   * 是否自动开始动画
   * @default true
   */
  autoStart?: boolean;
  /**
   * 文本翻滚方向，值为 down 和 up
   * @default down
   */
  direction?: RollingTextDirection;
  /**
   * 各个数位动画停止先后顺序，值为 ltr 和 rtl
   * @default ltr
   */
  stopOrder?: RollingTextStopOrder;
}

export type RollingTextInstance = {
  start: () => void;
  reset: () => void;
};

export interface RollingTextItemProps {
  figureArr?: any[];
  /**
   * 起始数值
   * @default 0
   */
  delay?: number;
  /**
   * 动画时长，单位为秒
   * @default 2
   */
  duration?: number;
  /**
   * 是否自动开始动画
   * @default true
   */
  isStart: boolean;
  /**
   * 文本翻滚方向，值为 down 和 up
   * @default down
   */
  direction?: RollingTextDirection;
  /**
   * 数字高度，单位为 px
   * @default 40
   */
  height?: number;
}

export type RollingTextThemeVars = {
  rollingTextBackground?: string;
  rollingTextColor?: string;
  rollingTextFontSize?: string;
  rollingTextGap?: string;
  rollingTextItemWidth?: string;
  rollingTextItemBorderRadius?: string;
};
