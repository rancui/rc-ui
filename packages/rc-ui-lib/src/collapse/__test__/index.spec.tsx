import * as React from 'react';
import { cleanup, act, render, fireEvent } from '@testing-library/react';
import Collapse from '..';
import { CollapseItemInstance } from '../PropsType';

describe('Collapse', () => {
  let spyConsole: jest.SpyInstance;
  beforeEach(() => {
    spyConsole = jest.spyOn(console, 'error');
    spyConsole.mockImplementation(() => {
      return null;
    });
  });

  afterEach(() => {
    cleanup();
    spyConsole.mockRestore();
  });

  // it('should update value when title is clicked', async () => {
  //   const onChange = jest.fn();
  //   const { container } = render(
  //     <Collapse value={[]} initValue="1" onChange={onChange}>
  //       <Collapse.Item title="标题1" name="first">
  //         内容
  //       </Collapse.Item>
  //       <Collapse.Item title="标题2" name="2">
  //         内容
  //       </Collapse.Item>
  //     </Collapse>,
  //   );
  //   const title = container.querySelector(Cell).children('.rc-collapse-item__title');
  //   await title.simulate('click');
  //   expect(onChange).toHaveBeenCalledWith(['first']);
  // });

  it('should update value when title is clicked in accordion mode', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Collapse value="" initValue="1" accordion onChange={onChange}>
        <Collapse.Item title="标题1" name="first">
          内容
        </Collapse.Item>
        <Collapse.Item title="标题2" name="2">
          内容
        </Collapse.Item>
      </Collapse>,
    );
    const title = container.querySelector('.rc-cell.rc-collapse-item__title');
    await fireEvent.click(title);
    expect(onChange).toHaveBeenCalledWith('first');
  });

  it('should render CollapseItem correctly', () => {
    const { container } = render(
      <Collapse value={[]}>
        <Collapse.Item
          title="标题1"
          name="first"
          icon="shop-o"
          value="this is value"
          rightIcon="question-o"
        >
          内容
        </Collapse.Item>
      </Collapse>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should not render border when border prop is false', () => {
    const { container } = render(<Collapse border={false} />);
    expect(container.querySelector('.rc-hairline--top-bottom')).toBeFalsy();
  });

  it('should toggle collapse after calling the toggle method', async () => {
    const CollapseItemRef = React.createRef<CollapseItemInstance>();
    const onChange = jest.fn();

    const toggle = (value?: boolean) => {
      CollapseItemRef.current.toggle(value);
    };

    render(
      <Collapse initValue={[]} onChange={onChange}>
        <Collapse.Item title="标题1" name="first" ref={CollapseItemRef}>
          内容
        </Collapse.Item>
      </Collapse>,
    );

    await act(async () => {
      toggle(true);
    });

    expect(onChange).toHaveBeenCalledWith(['first']);

    await act(async () => {
      toggle(false);
    });

    expect(onChange).toHaveBeenCalledWith([]);
  });

  it('should toggle collapse after calling the toggle method in accordion mode', async () => {
    const CollapseItemRef = React.createRef<CollapseItemInstance>();
    const onChange = jest.fn();

    const toggle = (value?: boolean) => {
      CollapseItemRef.current.toggle(value);
    };

    render(
      <Collapse initValue="" accordion onChange={onChange}>
        <Collapse.Item title="标题1" name="first" ref={CollapseItemRef}>
          内容
        </Collapse.Item>
      </Collapse>,
    );

    await act(async () => {
      toggle(true);
    });

    expect(onChange).toHaveBeenCalledWith('first');

    await act(async () => {
      toggle(false);
    });

    expect(onChange).toHaveBeenCalledWith('');
  });

  it('should be readonly when using readonly prop', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Collapse initValue="" onChange={onChange}>
        <Collapse.Item title="标题1" name="first" readonly>
          内容
        </Collapse.Item>
      </Collapse>,
    );
    const title = container.querySelector('.rc-cell.rc-collapse-item__title');
    await fireEvent.click(title);
    expect(onChange).not.toHaveBeenCalled();
    expect(container.querySelector('i')).toBeFalsy();
    expect(spyConsole).toHaveBeenCalledWith(
      '[rc-ui-lib] Collapse: "value" should be Array in non-accordion mode',
    );
  });

  it('should be readonly when using disabled prop', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Collapse initValue="" onChange={onChange}>
        <Collapse.Item title="标题1" name="first" disabled>
          内容
        </Collapse.Item>
      </Collapse>,
    );
    const title = container.querySelector('.rc-cell.rc-collapse-item__title');
    await fireEvent.click(title);
    expect(onChange).not.toHaveBeenCalled();
    expect(container.querySelector('.rc-collapse-item__title--disabled')).toBeTruthy();
    expect(spyConsole).toHaveBeenCalledWith(
      '[rc-ui-lib] Collapse: "value" should be Array in non-accordion mode',
    );
  });

  it('when value is change', async () => {
    const { container, rerender } = render(
      <Collapse initValue="" accordion value="">
        <Collapse.Item title="标题1" name="first">
          内容
        </Collapse.Item>
      </Collapse>,
    );

    const title = container.querySelector('.rc-cell.rc-collapse-item__title');
    await fireEvent.click(title);

    rerender(
      <Collapse initValue="" accordion value="first">
        <Collapse.Item title="标题1" name="first">
          内容
        </Collapse.Item>
      </Collapse>,
    );
    // expect(wrapper.props().value).toEqual('first');
    expect(container).toMatchSnapshot();
  });
});
