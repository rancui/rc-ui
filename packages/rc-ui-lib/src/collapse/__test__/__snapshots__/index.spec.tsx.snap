// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse should render CollapseItem correctly 1`] = `
<div>
  <div
    class="rc-collapse rc-hairline--top-bottom"
  >
    <div
      class="rc-collapse-item"
    >
      <div
        class="rc-cell rc-cell--clickable rc-collapse-item__title"
        role="button"
      >
        <i
          class="rc-badge__wrapper rc-cell__left-icon van-icon van-icon-shop-o"
        />
        <div
          class="rc-cell__title"
        >
          标题1
        </div>
        <div
          class="rc-cell__value"
        >
          <span>
            this is value
          </span>
        </div>
        question-o
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse when value is change 1`] = `
<div>
  <div
    class="rc-collapse rc-hairline--top-bottom"
  >
    <div
      class="rc-collapse-item"
    >
      <div
        class="rc-cell rc-cell--clickable rc-collapse-item__title rc-collapse-item__title--expanded"
        role="button"
      >
        <div
          class="rc-cell__title"
        >
          标题1
        </div>
        <i
          class="rc-badge__wrapper rc-cell__right-icon van-icon van-icon-arrow"
        />
      </div>
      <div
        class="rc-collapse-item__wrapper"
      >
        <div
          class="rc-collapse-item__content"
        >
          内容
        </div>
      </div>
    </div>
  </div>
</div>
`;
